File: app/backend/app/core/job_runner.py
"""
Manages long-running iterate jobs. Each job runs iterate_async and tracks
progress. Results are stored in an in-memory dict keyed by job_id.
"""

import asyncio
import uuid
from typing import Callable, Awaitable

from app.models.job import JobResult, JobStatus
from app.core import se_manager

_jobs: dict[str, JobResult] = {}


def get_job(job_id: str) -> JobResult | None:
    return _jobs.get(job_id)


def cancel_job(job_id: str) -> bool:
    job = _jobs.get(job_id)
    if job and job.status in (JobStatus.queued, JobStatus.running):
        se_manager.cancel_current()
        job.status = JobStatus.cancelled
        return True
    return False


async def submit_job(
    session_id: str,
    steps: int,
    progress_cb: Callable[[int, int, float], Awaitable[None]] | None = None,
) -> JobResult:
    job_id = str(uuid.uuid4())
    job = JobResult(
        job_id=job_id,
        session_id=session_id,
        status=JobStatus.queued,
        steps_requested=steps,
    )
    _jobs[job_id] = job

    async def _run() -> None:
        # Import here to avoid module-level circular import
        from app.ws.progress import broadcast

        job.status = JobStatus.running
        se_manager.clear_cancel()
        try:
            result = await se_manager.iterate_async(session_id, steps, progress_cb or _noop_cb)
            job.steps_completed = result["steps_completed"]
            job.energy_start = result.get("energy_start")
            job.energy_end = result.get("energy_end")
            job.status = JobStatus.completed
            await broadcast(session_id, {"type": "completed", "energy": job.energy_end})
        except asyncio.CancelledError:
            job.status = JobStatus.cancelled
        except Exception as exc:
            job.status = JobStatus.failed
            job.error = str(exc)
            await broadcast(session_id, {"type": "error", "message": str(exc)})

    asyncio.create_task(_run())
    return job


async def _noop_cb(step: int, total: int, energy: float) -> None:
    pass

File: app/backend-bun/src/job-runner.ts
import * as seManager from "./se-manager";
import * as wsHub from "./ws-hub";

export type JobStatus = "queued" | "running" | "completed" | "cancelled" | "failed";

export interface JobResult {
  job_id:          string;
  session_id:      string;
  status:          JobStatus;
  steps_requested: number;
  steps_completed: number;
  energy_start:    number | null;
  energy_end:      number | null;
  error:           string | null;
}

const jobs = new Map<string, JobResult>();

export function getJob(jobId: string): JobResult | undefined {
  return jobs.get(jobId);
}

export function cancelJob(jobId: string): boolean {
  const job = jobs.get(jobId);
  if (!job || (job.status !== "queued" && job.status !== "running")) return false;
  seManager.cancelCurrent();
  job.status = "cancelled";
  return true;
}

export async function submitJob(
  sessionId: string,
  steps: number,
  progressCb: (step: number, total: number, energy: number) => void,
): Promise<JobResult> {
  const jobId = crypto.randomUUID();
  const job: JobResult = {
    job_id:          jobId,
    session_id:      sessionId,
    status:          "queued",
    steps_requested: steps,
    steps_completed: 0,
    energy_start:    null,
    energy_end:      null,
    error:           null,
  };
  jobs.set(jobId, job);

  (async () => {
    job.status = "running";
    try {
      const result = await seManager.iterateAsync(sessionId, steps, async (step, total, energy) => {
        progressCb(step, total, energy);
      });
      job.steps_completed = result.steps_completed;
      job.energy_start    = result.energy_start;
      job.energy_end      = result.energy_end;
      if (job.status !== "cancelled") {
        job.status = "completed";
        wsHub.broadcast(sessionId, { type: "completed", energy: job.energy_end });
      }
    } catch (e) {
      if (job.status !== "cancelled") {
        job.status = "failed";
        job.error  = String(e);
        wsHub.broadcast(sessionId, { type: "error", message: String(e) });
      }
    }
  })();

  return job;
}
